File: apps/dsa-web/src/api/strategyLab.ts
import apiClient from './index';

export type StrategyLabStrategyItem = {
  strategy_id: string;
  name: string;
  instrument_types: string[];
  markets: string[];
  description?: string | null;
};

export type StrategyLabMetric = {
  total_return_pct?: number | null;
  annualized_return_pct?: number | null;
  max_drawdown_pct?: number | null;
  sharpe_ratio?: number | null;
  win_rate_pct?: number | null;
  trade_count: number;
  diagnostics?: Record<string, unknown>;
};

export type StrategyLabRunItem = {
  id: number;
  run_uid: string;
  strategy_id: string;
  strategy_name: string;
  engine_name: string;
  status: string;
  market: string;
  instrument_type: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  final_equity?: number | null;
  benchmark_symbol?: string | null;
  benchmark_return_pct?: number | null;
  portfolio_account_id?: number | null;
  error_message?: string | null;
  created_at?: string | null;
  completed_at?: string | null;
  parameters?: Record<string, unknown>;
  symbols?: string[];
  metrics?: StrategyLabMetric | null;
  equity_curve?: Array<{ trade_date: string; equity: number; cash: number; positions_value: number }>;
};

export type StrategyLabTradeItem = {
  id: number;
  run_id: number;
  trade_date: string;
  symbol: string;
  side: 'buy' | 'sell';
  quantity: number;
  price: number;
  amount: number;
  fee: number;
  reason?: string | null;
};

export type StrategyLabBatchItem = {
  id: number;
  batch_uid: string;
  strategy_id: string;
  strategy_name: string;
  market: string;
  instrument_type: string;
  status: string;
  total_tasks: number;
  completed_tasks: number;
  success_tasks: number;
  failed_tasks: number;
  created_at?: string | null;
  completed_at?: string | null;
  parameters_grid?: Array<Record<string, unknown>>;
  summary?: Record<string, number>;
  items?: Array<{ id: number; run_id?: number | null; parameters: Record<string, unknown>; status: string; error_message?: string | null }>;
};

export type StrategyLabSignalItem = {
  id: number;
  run_id: number;
  portfolio_account_id?: number | null;
  symbol: string;
  market: string;
  suggested_action: string;
  confidence?: number | null;
  reason?: string | null;
  status: string;
  portfolio_trade_id?: number | null;
  created_at?: string | null;
};

export type StrategyLabSyncRunItem = {
  id: number;
  sync_type: string;
  market: string;
  status: string;
  result: Record<string, number>;
  error_message?: string | null;
  created_at?: string | null;
};

export type StrategyLabEventStudyResponse = {
  market: string;
  event_type?: string | null;
  offsets: number[];
  total: number;
  summary: Record<string, { count: number; average_return_pct?: number | null }>;
  items: Array<{ bond_code: string; bond_name: string; event_date: string; event_type: string; returns_pct: Record<string, number | null> }>;
};

export type StrategyLabInstrumentItem = {
  bond_code: string;
  bond_name: string;
  stock_code: string;
  stock_name?: string | null;
  market: string;
  list_date?: string | null;
  maturity_date?: string | null;
  status?: string | null;
  remaining_size?: number | null;
  current_premium_rate?: number | null;
  convert_price?: number | null;
  latest_close?: number | null;
  latest_premium_rate?: number | null;
  event_count: number;
  source?: string | null;
  updated_at?: string | null;
};

export type StrategyLabInstrumentDetail = StrategyLabInstrumentItem & {
  industry?: string | null;
  terms: Record<string, unknown>;
  redeem_clause?: string | null;
  down_revise_clause?: string | null;
  put_clause?: string | null;
  redeem_trigger_price?: number | null;
  down_revise_trigger_price?: number | null;
  put_trigger_price?: number | null;
  bar_count: number;
};

export type StrategyLabBarItem = {
  trade_date: string;
  close?: number | null;
  premium_rate?: number | null;
  remaining_size?: number | null;
  redeem_alert: boolean;
  down_revise_alert: boolean;
  put_alert: boolean;
  source?: string | null;
};

export type StrategyLabEventItem = {
  event_date: string;
  event_type: string;
  event_detail?: string | null;
  source?: string | null;
  created_at?: string | null;
};

export type StrategyLabRunRequest = {
  strategy_id: string;
  market: string;
  instrument_type: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  benchmark_symbol?: string;
  symbols?: string[];
  parameters?: Record<string, unknown>;
};

export type StrategyLabBatchRequest = {
  strategy_id: string;
  market: string;
  instrument_type: string;
  base_config: Record<string, unknown>;
  parameter_grid: Record<string, unknown[]>;
  run_async?: boolean;
};

export const strategyLabApi = {
  async listStrategies() {
    const { data } = await apiClient.get<{ items: StrategyLabStrategyItem[] }>('/api/v1/strategy-lab/strategies');
    return data.items;
  },
  async listRuns() {
    const { data } = await apiClient.get<{ items: StrategyLabRunItem[] }>('/api/v1/strategy-lab/runs');
    return data.items;
  },
  async getRun(runId: number) {
    const { data } = await apiClient.get<StrategyLabRunItem>(`/api/v1/strategy-lab/runs/${runId}`);
    return data;
  },
  async listRunTrades(runId: number) {
    const { data } = await apiClient.get<{ items: StrategyLabTradeItem[] }>(`/api/v1/strategy-lab/runs/${runId}/trades`);
    return data.items;
  },
  async createRun(payload: StrategyLabRunRequest) {
    const { data } = await apiClient.post<StrategyLabRunItem>('/api/v1/strategy-lab/runs', payload);
    return data;
  },
  async listBatches() {
    const { data } = await apiClient.get<{ items: StrategyLabBatchItem[] }>('/api/v1/strategy-lab/batches');
    return data.items;
  },
  async getBatch(batchId: number) {
    const { data } = await apiClient.get<StrategyLabBatchItem>(`/api/v1/strategy-lab/batches/${batchId}`);
    return data;
  },
  async createBatch(payload: StrategyLabBatchRequest) {
    const { data } = await apiClient.post<StrategyLabBatchItem>('/api/v1/strategy-lab/batches', payload);
    return data;
  },
  async retryBatch(batchId: number) {
    const { data } = await apiClient.post<StrategyLabBatchItem>(`/api/v1/strategy-lab/batches/${batchId}/retry`);
    return data;
  },
  async resumeBatch(batchId: number) {
    const { data } = await apiClient.post<StrategyLabBatchItem>(`/api/v1/strategy-lab/batches/${batchId}/resume`);
    return data;
  },
  async deleteBatch(batchId: number) {
    await apiClient.delete(`/api/v1/strategy-lab/batches/${batchId}`);
  },
  getBatchStreamUrl(batchId: number) {
    const baseUrl = apiClient.defaults.baseURL || '';
    return `${baseUrl}/api/v1/strategy-lab/batches/${batchId}/stream`;
  },
  async listSignals() {
    const { data } = await apiClient.get<{ items: StrategyLabSignalItem[] }>('/api/v1/strategy-lab/signals');
    return data.items;
  },
  async createSignal(payload: { run_id: number; portfolio_account_id?: number; suggested_action: string; confidence?: number; reason?: string }) {
    const { data } = await apiClient.post<StrategyLabSignalItem>('/api/v1/strategy-lab/signals', payload);
    return data;
  },
  async confirmSignal(signalId: number, payload: { portfolio_account_id: number; trade_date: string; quantity: number; price: number; side: 'buy' | 'sell'; fee?: number; tax?: number }) {
    const { data } = await apiClient.post<StrategyLabSignalItem>(`/api/v1/strategy-lab/signals/${signalId}/confirm`, payload);
    return data;
  },
  async listSyncRuns(params: { page?: number; limit?: number } = {}) {
    const { data } = await apiClient.get<{ items: StrategyLabSyncRunItem[]; total: number }>('/api/v1/strategy-lab/data-sync/runs', { params });
    return data;
  },
  async syncData(payload: {
    market: string;
    source: string;
    sync_type?: string;
    include_delisted?: boolean;
    start_date?: string;
    end_date?: string;
    symbols?: string[];
  }) {
    const { data } = await apiClient.post('/api/v1/strategy-lab/data-sync', payload);
    return data;
  },
  async studyEvents(payload: { market: string; event_type?: string; offsets: number[]; symbols?: string[] }) {
    const { data } = await apiClient.post<StrategyLabEventStudyResponse>('/api/v1/strategy-lab/studies/events', payload);
    return data;
  },
  async listInstruments(params: { market?: string; keyword?: string; status?: 'active' | 'delisted'; held_only?: boolean; page?: number; limit?: number } = {}) {
    const { data } = await apiClient.get<{ market: string; total: number; page: number; limit: number; items: StrategyLabInstrumentItem[] }>('/api/v1/strategy-lab/instruments', { params });
    return data;
  },
  async getInstrumentDetail(bondCode: string, market = 'cn') {
    const { data } = await apiClient.get<StrategyLabInstrumentDetail>(`/api/v1/strategy-lab/instruments/${encodeURIComponent(bondCode)}`, { params: { market } });
    return data;
  },
  async listInstrumentBars(bondCode: string, params: { market?: string; start_date?: string; end_date?: string; limit?: number } = {}) {
    const { data } = await apiClient.get<{ bond_code: string; total: number; items: StrategyLabBarItem[] }>(`/api/v1/strategy-lab/instruments/${encodeURIComponent(bondCode)}/bars`, { params });
    return data;
  },
  async listInstrumentEvents(bondCode: string, params: { market?: string; event_type?: string; limit?: number } = {}) {
    const { data } = await apiClient.get<{ bond_code: string; total: number; items: StrategyLabEventItem[] }>(`/api/v1/strategy-lab/instruments/${encodeURIComponent(bondCode)}/events`, { params });
    return data;
  },
};

File: api/v1/schemas/strategy_lab.py
# -*- coding: utf-8 -*-
"""Strategy Lab API schemas."""

from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


StrategyLabMarket = Literal["cn", "hk", "us"]
StrategyLabInstrumentType = Literal["convertible_bond", "stock", "hk_stock", "us_stock"]
StrategyLabRunStatus = Literal["pending", "running", "completed", "failed"]
StrategyLabTradeSide = Literal["buy", "sell"]


class StrategyLabStrategyItem(BaseModel):
    strategy_id: str
    name: str
    instrument_types: List[str] = Field(default_factory=list)
    markets: List[str] = Field(default_factory=list)
    description: Optional[str] = None


class StrategyLabStrategyListResponse(BaseModel):
    items: List[StrategyLabStrategyItem] = Field(default_factory=list)


class StrategyLabRunCreateRequest(BaseModel):
    strategy_id: str = Field("double-low", description="策略 ID")
    market: StrategyLabMarket = Field("cn", description="市场")
    instrument_type: StrategyLabInstrumentType = Field("convertible_bond", description="品种类型")
    start_date: date = Field(..., description="回测起始日期")
    end_date: date = Field(..., description="回测结束日期")
    initial_cash: float = Field(100000.0, gt=0, description="初始资金")
    benchmark_symbol: Optional[str] = Field(None, description="基准标的")
    symbols: List[str] = Field(default_factory=list, description="可选标的筛选")
    parameters: Dict[str, Any] = Field(default_factory=dict, description="策略参数")
    portfolio_account_id: Optional[int] = Field(None, description="关联的 Portfolio 账户 ID")


class StrategyLabMetricItem(BaseModel):
    total_return_pct: Optional[float] = None
    annualized_return_pct: Optional[float] = None
    max_drawdown_pct: Optional[float] = None
    sharpe_ratio: Optional[float] = None
    sortino_ratio: Optional[float] = None
    calmar_ratio: Optional[float] = None
    win_rate_pct: Optional[float] = None
    trade_count: int = 0
    exposure_days: int = 0
    diagnostics: Dict[str, Any] = Field(default_factory=dict)


class StrategyLabEquityPointItem(BaseModel):
    trade_date: str
    equity: float
    cash: float
    positions_value: float


class StrategyLabRunSummaryItem(BaseModel):
    id: int
    run_uid: str
    strategy_id: str
    strategy_name: str
    engine_name: str
    status: StrategyLabRunStatus
    market: str
    instrument_type: str
    start_date: str
    end_date: str
    initial_cash: float
    final_equity: Optional[float] = None
    benchmark_symbol: Optional[str] = None
    benchmark_return_pct: Optional[float] = None
    portfolio_account_id: Optional[int] = None
    error_message: Optional[str] = None
    created_at: Optional[str] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None


class StrategyLabRunItem(StrategyLabRunSummaryItem):
    parameters: Dict[str, Any] = Field(default_factory=dict)
    symbols: List[str] = Field(default_factory=list)
    metrics: Optional[StrategyLabMetricItem] = None
    equity_curve: List[StrategyLabEquityPointItem] = Field(default_factory=list)


class StrategyLabRunListResponse(BaseModel):
    total: int
    page: int
    limit: int
    items: List[StrategyLabRunSummaryItem] = Field(default_factory=list)


class StrategyLabTradeItem(BaseModel):
    id: int
    run_id: int
    trade_date: str
    canonical_id: str
    symbol: str
    market: str
    instrument_type: str
    side: StrategyLabTradeSide
    quantity: float
    price: float
    amount: float
    fee: float = 0.0
    reason: Optional[str] = None
    portfolio_trade_id: Optional[int] = None


class StrategyLabTradeListResponse(BaseModel):
    run_id: int
    items: List[StrategyLabTradeItem] = Field(default_factory=list)


class StrategyLabDataSyncRequest(BaseModel):
    market: StrategyLabMarket = Field("cn", description="市场")
    source: str = Field("opencli", description="同步来源")
    sync_type: str = Field("", description="opencli 同步类型：cb_basic / cb_ohlc / cb_premium_history / all")
    include_delisted: bool = Field(False, description="是否同步已退市可转债（默认仅活跃）")
    start_date: Optional[date] = Field(None, description="行情同步起始日期（缺省时增量）")
    end_date: Optional[date] = Field(None, description="行情同步结束日期（默认今天）")
    symbols: List[str] = Field(default_factory=list, description="可选标的筛选")


class StrategyLabDataSyncResponse(BaseModel):
    sync_run_id: int = 0
    status: str = "running"
    sync_type: str = ""
    cb_basic_upserted: int = 0
    cb_terms_upserted: int = 0
    cb_factor_upserted: int = 0
    cb_event_upserted: int = 0
    ohlc_bars_upserted: int = 0
    ohlc_skipped: int = 0
    cb_factor_rows_patched: int = 0
    premium_rate_patched: int = 0
    remaining_size_patched: int = 0
    failed_bonds: List[Any] = Field(default_factory=list)


class StrategyLabSyncRunItem(BaseModel):
    id: int
    run_uid: str
    sync_type: str
    market: str
    status: str
    cancel_requested: bool = False
    result: Dict[str, Any] = Field(default_factory=dict)
    error_message: Optional[str] = None
    created_at: Optional[str] = None
    completed_at: Optional[str] = None


class StrategyLabSyncRunListResponse(BaseModel):
    page: int
    limit: int
    total: int
    items: List[StrategyLabSyncRunItem] = Field(default_factory=list)


class StrategyLabInstrumentItem(BaseModel):
    bond_code: str
    bond_name: str
    stock_code: str
    stock_name: Optional[str] = None
    market: str
    list_date: Optional[str] = None
    maturity_date: Optional[str] = None
    status: Optional[str] = None
    remaining_size: Optional[float] = None
    current_premium_rate: Optional[float] = None
    convert_price: Optional[float] = None
    latest_close: Optional[float] = None
    latest_premium_rate: Optional[float] = None
    event_count: int = 0
    source: Optional[str] = None
    updated_at: Optional[str] = None


class StrategyLabInstrumentListResponse(BaseModel):
    market: str
    total: int
    page: int
    limit: int
    items: List[StrategyLabInstrumentItem] = Field(default_factory=list)


class StrategyLabInstrumentDetailItem(BaseModel):
    bond_code: str
    bond_name: str
    stock_code: str
    stock_name: Optional[str] = None
    market: str
    list_date: Optional[str] = None
    maturity_date: Optional[str] = None
    status: Optional[str] = None
    remaining_size: Optional[float] = None
    current_premium_rate: Optional[float] = None
    convert_price: Optional[float] = None
    latest_close: Optional[float] = None
    latest_premium_rate: Optional[float] = None
    industry: Optional[str] = None
    terms: Dict[str, Any] = Field(default_factory=dict)
    redeem_clause: Optional[str] = None
    down_revise_clause: Optional[str] = None
    put_clause: Optional[str] = None
    redeem_trigger_price: Optional[float] = None
    down_revise_trigger_price: Optional[float] = None
    put_trigger_price: Optional[float] = None
    source: Optional[str] = None
    updated_at: Optional[str] = None
    bar_count: int = 0
    event_count: int = 0


class StrategyLabBarItem(BaseModel):
    trade_date: Optional[str] = None
    close: Optional[float] = None
    premium_rate: Optional[float] = None
    remaining_size: Optional[float] = None
    redeem_alert: bool = False
    down_revise_alert: bool = False
    put_alert: bool = False
    source: Optional[str] = None


class StrategyLabBarListResponse(BaseModel):
    bond_code: str
    total: int
    items: List[StrategyLabBarItem] = Field(default_factory=list)


class StrategyLabEventItem(BaseModel):
    event_date: str
    event_type: str
    event_detail: Optional[str] = None
    source: Optional[str] = None
    created_at: Optional[str] = None


class StrategyLabEventListResponse(BaseModel):
    bond_code: str
    total: int
    items: List[StrategyLabEventItem] = Field(default_factory=list)


class StrategyLabEventStudyRequest(BaseModel):
    market: StrategyLabMarket = Field("cn", description="市场")
    event_type: Optional[str] = Field(None, description="可选事件类型筛选")
    offsets: List[int] = Field(default_factory=lambda: [-5, -1, 1, 5], description="相对事件日的交易日偏移")
    symbols: List[str] = Field(default_factory=list, description="可选可转债代码筛选")


class StrategyLabEventStudyItem(BaseModel):
    bond_code: str
    bond_name: str
    event_date: str
    event_type: str
    base_trade_date: str
    base_close: float
    returns_pct: Dict[str, Optional[float]] = Field(default_factory=dict)


class StrategyLabEventStudyResponse(BaseModel):
    market: str
    event_type: Optional[str] = None
    offsets: List[int]
    total: int
    summary: Dict[str, Dict[str, Optional[float] | int]] = Field(default_factory=dict)
    items: List[StrategyLabEventStudyItem] = Field(default_factory=list)


class StrategyLabBatchCreateRequest(BaseModel):
    strategy_id: str = Field("double-low", description="策略 ID")
    market: StrategyLabMarket = Field("cn", description="市场")
    instrument_type: StrategyLabInstrumentType = Field("convertible_bond", description="品种类型")
    base_config: Dict[str, Any] = Field(default_factory=dict, description="基础运行参数")
    parameter_grid: Dict[str, List[Any]] = Field(default_factory=dict, description="参数网格")
    run_async: bool = Field(False, description="是否交给后台任务执行并通过 SSE 观察进度")


class StrategyLabBatchItemItem(BaseModel):
    id: int
    batch_id: int
    run_id: Optional[int] = None
    parameters: Dict[str, Any] = Field(default_factory=dict)
    status: str
    error_message: Optional[str] = None
    created_at: Optional[str] = None
    completed_at: Optional[str] = None


class StrategyLabBatchSummaryItem(BaseModel):
    id: int
    batch_uid: str
    strategy_id: str
    strategy_name: str
    market: str
    instrument_type: str
    status: str
    total_tasks: int
    completed_tasks: int
    success_tasks: int
    failed_tasks: int
    created_at: Optional[str] = None
    completed_at: Optional[str] = None


class StrategyLabBatchItemResponse(StrategyLabBatchSummaryItem):
    parameters_grid: List[Dict[str, Any]] = Field(default_factory=list)
    summary: Dict[str, Any] = Field(default_factory=dict)
    items: List[StrategyLabBatchItemItem] = Field(default_factory=list)


class StrategyLabBatchListResponse(BaseModel):
    page: int
    limit: int
    total: int
    items: List[StrategyLabBatchSummaryItem] = Field(default_factory=list)


class StrategyLabSignalCreateRequest(BaseModel):
    run_id: int
    portfolio_account_id: Optional[int] = None
    suggested_action: str = Field("hold", description="建议动作")
    signal_type: str = Field("strategy_recommendation", description="信号类型")
    confidence: Optional[float] = Field(None, ge=0, le=1, description="置信度")
    reason: Optional[str] = None


class StrategyLabSignalItem(BaseModel):
    id: int
    run_id: int
    portfolio_account_id: Optional[int] = None
    canonical_id: str
    symbol: str
    market: str
    instrument_type: str
    signal_type: str
    suggested_action: str
    confidence: Optional[float] = None
    reason: Optional[str] = None
    status: str
    portfolio_trade_id: Optional[int] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class StrategyLabSignalListResponse(BaseModel):
    page: int
    limit: int
    total: int
    items: List[StrategyLabSignalItem] = Field(default_factory=list)


class StrategyLabSignalConfirmRequest(BaseModel):
    portfolio_account_id: int
    trade_date: date
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    side: Optional[StrategyLabTradeSide] = None
    fee: float = Field(0.0, ge=0)
    tax: float = Field(0.0, ge=0)
